docs/toctoc/guides/token-based-authentication-flow.md

```diff
@@ -0,0 +1,210 @@
+  authenticated `Subject` if true.
+
+- `unregisterToken`: invalidates an `AccessToken`. This is typically used when a
+  user logs out from a device.
+
+- `unregisterAllSubjectTokens`: invalidates all `AccessToken` of a `Subject`.
+  This can be used to implement security features, such as logging out from all
+  active sessions in case the password is suspected to be compromised or upon a
+  change password request.
+
+- `unregisterLogin`: invalidates a `Login`. This can be used in case of a change
+  password request in an app that wants to guarantee a single login at the time
+  per user.
+
+- `unregisterAllSubjectLogins`: invalidates all `Login` of a `Subject`. This is
+  typically used when removing a user from the system.
+
+## How to use it
+
+You can create an `TokenBasedAuthenticationFlow` by providing the two
+authentication domains discussed above.
+
+`toctoc` provides some specialized implementations of such domains combinations
+of DBMS and data-access library. For this example we'll be using Postgres and
+Slick, leveraging the `toctoc-slick-postgresql` module (see the
+[installation docs](toctoc/installation.md) for instructions on how to add it as
+a dependency).
+
+```scala mdoc
+// The generic flow
+import io.buildo.toctoc.core.authentication.TokenBasedAuthentication.TokenBasedAuthenticationFlow
+
+// The specialized AccessToken domain
+import io.buildo.toctoc.slick.authentication.token.PostgreSqlSlickAccessTokenAuthenticationDomain
+
+// The specialized Login domain
+import io.buildo.toctoc.slick.authentication.login.PostgreSqlSlickLoginAuthenticationDomain
+
+import cats.effect.Sync
+import monix.catnap.FutureLift
+import slick.jdbc.JdbcBackend.Database
+
+import scala.concurrent.Future
+import java.time.Duration
+
+object PostgreSqlSlickTokenBasedAuthenticationFlow {
+  def create[F[_]: Sync: FutureLift[?[_], Future]](
+    db: Database,
+    tokenDuration: Duration
+  ): TokenBasedAuthenticationFlow[F] = new TokenBasedAuthenticationFlow[F](
+    loginD = new PostgreSqlSlickLoginAuthenticationDomain[F](db),
+    accessTokenD = new PostgreSqlSlickAccessTokenAuthenticationDomain[F](db),
+    tokenDuration = tokenDuration,
+  )
+}
+```
+
+Now that we have it, we can integrate it in our app.
+
+### Login and Logout
+
+Here's a possible implementation of a controller that provides `login` and
+`logout` functionalities:
+
+```scala mdoc
+import io.buildo.toctoc.core.authentication.TokenBasedAuthentication.TokenBasedAuthenticationFlow
+import io.buildo.toctoc.core.authentication.TokenBasedAuthentication.AccessToken
+import io.buildo.toctoc.core.authentication.TokenBasedAuthentication.Login
+import io.buildo.toctoc.core.authentication.AuthenticationError
+import cats.effect.Sync
+
+trait AuthenticationController[F[_]] {
+  def login(credentials: Login): F[Either[AuthenticationError, AccessToken]]
+  def logout(accessToken: AccessToken): F[Either[AuthenticationError, Unit]]
+}
+
+object AuthenticationController {
```

Comment:
Why are we using this sort of `builder pattern` in the example? I would make things easier. Possibly, I'd remove even the trait.

Reply:
That's the pattern we're using in TF applications for reasons that we can discuss offline. With that in mind, I wanted to keep the example as close as possible to production code.